src/machine/machine_nrf_rand_softdevice.go

```diff
@@ -0,0 +1,42 @@
+//go:build nrf && softdevice
+// +build nrf,softdevice
+
+package machine
+
+import (
+	"device/arm"
+	"device/nrf"
+)
+
+//export sd_rand_application_vector_get
+func sd_rand_application_vector_get(*[4]uint8, uint8)
+
+// This is a global variable to avoid a heap allocation in GetRNG.
+var softdeviceEnabled uint8
+var rngData [4]uint8
+
+// GetRNG returns 32 bits of cryptographically secure random data
+func GetRNG() (uint32, error) {
+	// First check whether the SoftDevice is enabled.
+	arm.SVCall1(0x12, &softdeviceEnabled) // sd_softdevice_is_enabled
+
+	if softdeviceEnabled != 0 {
+		// Now pick the appropriate SVCall number. Hopefully they won't change
+		// in the future with a different SoftDevice version.
+		if nrf.Device == "nrf51" {
+			// sd_rand_application_vector_get: SOC_SVC_BASE_NOT_AVAILABLE + 15
+			arm.SVCall2(0x2B+15, &rngData, 4)
+		} else if nrf.Device == "nrf52" || nrf.Device == "nrf52840" || nrf.Device == "nrf52833" {
+			// sd_rand_application_vector_get: SOC_SVC_BASE_NOT_AVAILABLE + 5
+			arm.SVCall2(0x2C+5, &rngData, 4)
+		} else {
+			sd_rand_application_vector_get(&rngData, 4)
+		}
```

Comment:
Unfortunately this isn't entirely complete. The call can also return `NRF_ERROR_SOC_RAND_NOT_ENOUGH_VALUES`, in which case no random bytes are returned.

Reply:
See latest commit for this error check.